File: packages/vaultsync/vaultsync/sync_recovery.py
"""Bounded admission fence and explicit recovery; no automatic fork selection."""

from __future__ import annotations

import base64
import hashlib
import json
import threading
from pathlib import Path

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey

from .authenticated_state_view import (
    EMPTY_REGISTRY,
    LIMIT,
    ZERO,
    StateViewError,
    _boundary,
    _reject,
    _verified,
    registry_root,
)
from .sync_queue import (
    _ROOT_SIGNATURE_DOMAIN,
    OpaqueCiphertextEnvelope,
    SignedStateCommitment,
    _canonical_json,
    _commitment_sequence,
    _EncryptedQueueFile,
    _identifier,
    _state_digest,
)

_PENDING = "ATLAS_RECOVERY_PENDING"
_META = ("sequence", "root", "registry_root", "key_epoch")


class GuardedSyncState:
    def __init__(
        self,
        path: Path,
        *,
        encryption_key: bytes,
        account_id: str,
        vault_id: str,
        collection_id: str,
        key_epoch: int,
        trusted_signer: bytes,
    ):
        with _boundary():
            if len(trusted_signer) != 32:
                _reject()
            self._public = bytes(trusted_signer)
            self._context = {
                "account_id": _identifier(account_id),
                "vault_id": _identifier(vault_id),
                "collection_id": _identifier(collection_id),
                "key_epoch": _commitment_sequence(key_epoch),
                "signing_public_b64": base64.b64encode(trusted_signer).decode("ascii"),
            }
            self._store = _EncryptedQueueFile(
                Path(path), encryption_key=encryption_key, kind="guarded-sync-state-v1"
            )
            self._lock = threading.RLock()

    def _chain(self, raw):
        if not isinstance(raw, list) or len(raw) > LIMIT:
            _reject("ATLAS_HISTORY_LIMIT")
        views, previous, registry = [], ZERO, EMPTY_REGISTRY
        for i, item in enumerate(raw):
            v = _verified(item, self._public)
            if any(v[k] != self._context[k] for k in ("account_id", "vault_id", "key_epoch")):
                _reject()
            if (
                v["sequence"] != i + 1
                or v["previous_root"] != previous
                or v["previous_registry_root"] != registry
            ):
                _reject()
            views.append(v)
            previous, registry = v["root"], v["registry_root"]
        return views

    def _load(self):
        s = self._store.read({})
        if (
            set(s) != {"context", "views", "records", "cases", "status"}
            or s["context"] != self._context
        ):
            _reject()
        s["views"] = self._chain(s["views"])
        if (
            s["status"] not in ("ACTIVE", "MANUAL_REQUIRED", "RECOVERY_PENDING")
            or not isinstance(s["cases"], list)
            or len(s["cases"]) > 8
        ):
            _reject()
        if not isinstance(s["records"], dict) or len(s["records"]) > LIMIT:
            _reject()
        return s

    def initialize(self):
        with self._lock, _boundary():
            if self._store.path.exists():
                _reject()
            self._store.write(
                {
                    "context": self._context,
                    "views": [],
                    "records": {},
                    "cases": [],
                    "status": "ACTIVE",
                }
            )

    def _active(self, s):
        if s["status"] != "ACTIVE":
            _reject(_PENDING)
        if len(s["cases"]) == 8:
            s["status"] = "RECOVERY_PENDING"
            self._store.write(s)
            _reject(_PENDING)

    def automatic_sync(self, operation):
        with self._lock, _boundary():
            self._active(self._load())
            return operation()

    def checkpoint(self):
        with self._lock, _boundary():
            s = self._load()
            return {
                "sequence": len(s["views"]),
                "cursor": s["views"][-1]["root"] if s["views"] else ZERO,
                "records": [s["records"][k] for k in sorted(s["records"])],
            }

    def export_evidence(self):
        with self._lock, _boundary():
            return self._load()["views"]

    def _alarm(self, s, reason, peer, registry=None):
        s["cases"].append(
            {
                "reason": reason,
                "local": list(s["views"]),
                "peer": peer,
                "presented_registry_root": registry,
                "disposition": None,
                "rejected_branch": None,
            }
        )
        s["status"] = "MANUAL_REQUIRED"
        self._store.write(s)
        _reject(reason)

    def evidence(self):
        with self._lock, _boundary():
            s = self._load()
            c = s["cases"][-1] if s["cases"] else {"local": s["views"], "peer": []}
            return {"local": c["local"], "peer": c["peer"]}

    def recovery(self):
        with self._lock, _boundary():
            s = self._load()
            c = s["cases"][-1] if s["cases"] else None
            return {
                "status": s["status"],
                "reason": c["reason"] if c else None,
                "local": [{k: v[k] for k in _META} for v in (c["local"] if c else s["views"])],
                "peer": [{k: v[k] for k in _META} for v in (c["peer"] if c else [])],
                "disposition": c["disposition"] if c else None,
                "rejected_branch": c["rejected_branch"] if c else None,
                "presented_registry_root": c["presented_registry_root"] if c else None,
            }

    def resolve(self, disposition, local_root, peer_root):
        with self._lock, _boundary():
            s = self._load()
            if s["status"] != "MANUAL_REQUIRED" or disposition not in (
                "retain_accepted",
                "select_peer",
                "keep_blocked",
            ):
                _reject()
            c = s["cases"][-1]
            if local_root != (c["local"][-1]["root"] if c["local"] else ZERO) or peer_root != (
                c["peer"][-1]["root"] if c["peer"] else ZERO
            ):
                _reject()
            # Only rejection of an already-known signed replay can safely resume v1.
            safe = (
                c["reason"] == "ATLAS_ROLLBACK_REJECTED"
                and bool(c["peer"])
                and all(
                    v["sequence"] <= len(c["local"])
                    and v["root"] == c["local"][v["sequence"] - 1]["root"]
                    for v in c["peer"]
                )
            )
            c["disposition"] = disposition
            c["rejected_branch"] = {
                "retain_accepted": "peer",
                "select_peer": "local",
                "keep_blocked": None,
            }[disposition]
            s["status"] = (
                "ACTIVE" if disposition == "retain_accepted" and safe else "RECOVERY_PENDING"
            )
            self._store.write(s)
            return s["status"]

    def compare_evidence(self, peer):
        with self._lock, _boundary():
            s = self._load()
            self._active(s)
            signed = []
            try:
                with _boundary():
                    if not isinstance(peer, list) or len(peer) > LIMIT:
                        _reject("ATLAS_HISTORY_LIMIT")
                    signed = [_verified(v, self._public) for v in peer]
                    checked = self._chain(signed)
                    if not checked or not s["views"]:
                        _reject("ATLAS_CHECKPOINT_REQUIRED")
                    if any(a["root"] != b["root"] for a, b in zip(s["views"], checked)):
                        _reject("ATLAS_STATE_EQUIVOCATION")
                    return min(len(s["views"]), len(checked))
            except StateViewError as e:
                self._alarm(s, str(e), signed)

    def ingest(self, raw_view, registry, raw_collection, opaque_state):
        with self._lock, _boundary():
            s = self._load()
            self._active(s)
            peer = []
            registry_digest = None
            try:
                with _boundary():
                    view = _verified(raw_view, self._public)
                    peer = [view]
                    if any(
                        view[k] != self._context[k] for k in ("account_id", "vault_id", "key_epoch")
                    ):
                        _reject()
                    registry_digest = registry_root(registry)
                    if registry_digest != view["registry_root"]:
                        _reject("ATLAS_REGISTRY_SUBSTITUTION")
                    if len(opaque_state) > 1024 * 1024:
                        _reject("ATLAS_HISTORY_LIMIT")
                    c = SignedStateCommitment.from_dict(raw_collection)
                    if (
                        c.collection_id != self._context["collection_id"]
                        or c.sequence != view["sequence"]
                        or c.root != view["collection_root"]
                        or c.state_sha256 != _state_digest(opaque_state)
                    ):
                        _reject()
                    Ed25519PublicKey.from_public_bytes(self._public).verify(
                        base64.b64decode(c.signature_b64),
                        _ROOT_SIGNATURE_DOMAIN + bytes.fromhex(c.root),
                    )
                    n = view["sequence"]
                    views = s["views"]
                    if n <= len(views) and view["root"] != views[n - 1]["root"]:
                        _reject("ATLAS_STATE_EQUIVOCATION")
                    if n < len(views):
                        _reject("ATLAS_ROLLBACK_REJECTED")
                    if n == len(views):
                        return False
                    if n > LIMIT:
                        _reject("ATLAS_HISTORY_LIMIT")
                    previous = views[-1] if views else None
                    if (
                        n != len(views) + 1
                        or view["previous_root"] != (previous["root"] if previous else ZERO)
                        or view["previous_registry_root"]
                        != (previous["registry_root"] if previous else EMPTY_REGISTRY)
                        or c.previous_root != (previous["collection_root"] if previous else ZERO)
                    ):
                        _reject()
                    if len(opaque_state) > 1024 * 1024:
                        _reject("ATLAS_HISTORY_LIMIT")
                    body = json.loads(opaque_state)
                    if (
                        set(body) != {"format", "version", "route", "records"}
                        or body["format"] != "atlasvault-guarded-collection"
                        or type(body["version"]) is not int
                        or body["version"] != 1
                        or body["route"] not in ("patch", "snapshot", "compaction")
                    ):
                        _reject()
                    if not isinstance(body["records"], list) or len(body["records"]) > LIMIT:
                        _reject("ATLAS_HISTORY_LIMIT")
                    records = {}
                    for raw in body["records"]:
                        r = OpaqueCiphertextEnvelope.from_dict(raw)
                        if (
                            r.version != 1
                            or r.object_id in records
                            or r.key_epoch != view["key_epoch"]
                        ):
                            _reject()
                        records[r.object_id] = {
                            "object_id": r.object_id,
                            "revision": r.revision,
                            "content_sha256": r.content_sha256,
                            "envelope_sha256": hashlib.sha256(
                                _canonical_json(r.to_dict())
                            ).hexdigest(),
                            "tombstone": r.tombstone,
                        }
                    for key, old in s["records"].items():
                        if old["tombstone"] and records.get(key) != old:
                            _reject("ATLAS_TOMBSTONE_RESURRECTION")
                        if key not in records:
                            _reject("ATLAS_STALE_STATE")
                    s["records"] = records
                    s["views"] = [*views, view]
            except StateViewError as e:
                self._alarm(s, str(e), peer, registry_digest)
            self._store.write(s)
            return True

File: apps/atlas_flutter/lib/src/atlas_vault/sync_recovery.dart
part of 'sync_queue.dart';

/// Atomic admission, terminal anchors and manual recovery, with one owner per file.
final class AtlasVaultGuardedSyncState {
  AtlasVaultGuardedSyncState({
    required File file,
    required Uint8List encryptionKey,
    required String accountId,
    required String vaultId,
    required String collectionId,
    required int keyEpoch,
    required Uint8List trustedSigner,
    this._rotationRegistry,
  }) : _store = _EncryptedQueueFile(
         file,
         encryptionKey,
         kind: 'guarded-sync-state-v1',
       ),
       _public = Uint8List.fromList(trustedSigner),
       _context = {
         'account_id': _commitmentIdentifier(accountId),
         'vault_id': _commitmentIdentifier(vaultId),
         'collection_id': _commitmentIdentifier(collectionId),
         'key_epoch': _commitmentSequence(keyEpoch),
         'signing_public_b64': base64Encode(trustedSigner),
       } {
    if (trustedSigner.length != 32) _viewFail();
  }
  _EncryptedQueueFile _store;
  final List<Map<String, Object?>>? _rotationRegistry;
  final Uint8List _public;
  final Map<String, Object?> _context;
  bool _busy = false;
  Future<T> _run<T>(Future<T> Function() operation) async {
    if (_busy) _viewFail();
    _busy = true;
    try {
      return await operation();
    } on AtlasVaultStateViewException {
      rethrow;
    } catch (_) {
      _viewFail();
    } finally {
      _busy = false;
    }
  }

  Future<T> _checked<T>(Future<T> Function() operation) async {
    try {
      return await operation();
    } on AtlasVaultStateViewException {
      rethrow;
    } catch (_) {
      _viewFail();
    }
  }

  Future<List<Map<String, Object?>>> _chain(
    List<Map<String, Object?>> raw, [
    List<Map<String, Object?>> proofs = const [],
  ]) async {
    if (raw.length > _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
    final views = <Map<String, Object?>>[];
    var previous = _zeroRoot, registry = _emptyRegistryRoot;
    var epoch = _context['key_epoch']! as int, public = _public;
    Map<String, Object?>? plan;
    for (var i = 0; i < raw.length; i++) {
      for (final proof in proofs) {
        final candidate = _object(proof['plan']);
        if (previous == candidate['state_root'] &&
            epoch == candidate['previous_epoch']) {
          plan = candidate;
          epoch = candidate['new_epoch'] as int;
          public = _bridgePublic(proof);
        }
      }
      final v = await AtlasVaultAuthenticatedStateView._verify(raw[i], public);
      if (['account_id', 'vault_id'].any((k) => v[k] != _context[k]) ||
          v['key_epoch'] != epoch) {
        _viewFail();
      }
      if (plan != null &&
          epoch == plan['new_epoch'] &&
          v['registry_root'] != plan['resulting_registry_root']) {
        _viewFail();
      }
      if (v['sequence'] != i + 1 ||
          v['previous_root'] != previous ||
          v['previous_registry_root'] != registry) {
        _viewFail();
      }
      views.add(v);
      previous = v['root']! as String;
      registry = v['registry_root']! as String;
    }
    return views;
  }

  List<Map<String, Object?>> _views(Object? value) =>
      (value! as List).map(_object).toList();
  Uint8List _bridgePublic(Map<String, Object?> proof) => base64Decode(
    _views(proof['registry']).firstWhere(
          (e) => e['device_id'] == proof['rotation_signer_device_id'],
        )['signing_public_b64']!
        as String,
  );
  Future<List<Map<String, Object?>>> _bridge(Map<String, Object?> state) async {
    final records = _epochBridgeRecords(state);
    if (records.isNotEmpty && _rotationRegistry == null) _viewFail();
    final proofs = await _verifyEpochBridges(
      records,
      _rotationRegistry ?? [],
      _context,
    );
    final roots = _views(state['views']).map((v) => v['root']).toList();
    var position = -1;
    for (final proof in proofs) {
      final found = roots.indexOf(_object(proof['plan'])['state_root']);
      if (found < 0 || found < position) _viewFail();
      position = found;
    }
    return proofs;
  }

  Future<Map<String, Object?>> _stageEpoch(Map<String, Object?> proof) async {
    final s = await _load();
    await _active(s);
    if (s['epoch_bridge'] != null ||
        _views(s['views']).isEmpty ||
        _views(s['views']).last['root'] !=
            _object(proof['plan'])['state_root']) {
      _viewFail('ATLAS_RECOVERY_PENDING');
    }
    s['epoch_bridge'] = proof;
    await _bridge(s);
    await _chain(_views(s['views']), await _bridge(s));
    return s;
  }

  Future<Map<String, Object?>> _load() async {
    final s = await _store.read({});
    _exact(s, {
      'context',
      'views',
      'records',
      'cases',
      'status',
      if (s.containsKey('epoch_bridge')) 'epoch_bridge',
      if (s.containsKey('epoch_bridges')) 'epoch_bridges',
    });
    final context = _object(s['context']);
    _exact(context, _context.keys.toSet());
    if (_context.entries.any((e) => context[e.key] != e.value) ||
        ![
          'ACTIVE',
          'MANUAL_REQUIRED',
          'RECOVERY_PENDING',
        ].contains(s['status']) ||
        (s['cases']! as List).length > 8 ||
        _object(s['records']).length > _viewLimit) {
      _viewFail();
    }
    s['views'] = await _chain(_views(s['views']), await _bridge(s));
    return s;
  }

  Future<void> initialize() => _run(() async {
    if (await _store.file.exists()) _viewFail();
    await _store.write({
      'context': _context,
      'views': [],
      'records': <String, Object?>{},
      'cases': [],
      'status': 'ACTIVE',
    });
  });
  Future<void> _active(Map<String, Object?> s) async {
    if (s['status'] != 'ACTIVE') _viewFail('ATLAS_RECOVERY_PENDING');
    if ((s['cases']! as List).length == 8) {
      s['status'] = 'RECOVERY_PENDING';
      await _store.write(s);
      _viewFail('ATLAS_RECOVERY_PENDING');
    }
  }

  Future<T> automaticSync<T>(Future<T> Function() operation) => _run(() async {
    await _active(await _load());
    return operation();
  });
  Future<Map<String, Object?>> checkpoint() => _run(() async {
    final s = await _load(),
        views = _views(s['views']),
        records = _object(s['records']),
        keys = records.keys.toList()..sort();
    return {
      'sequence': views.length,
      'cursor': views.isEmpty ? _zeroRoot : views.last['root'],
      'records': [for (final k in keys) records[k]],
    };
  });
  Future<List<Map<String, Object?>>> exportEvidence() =>
      _run(() async => _views((await _load())['views']));
  Future<Never> _alarm(
    Map<String, Object?> s,
    String reason,
    List<Map<String, Object?>> peer, [
    String? registry,
  ]) async {
    (s['cases']! as List).add({
      'reason': reason,
      'local': _views(s['views']),
      'peer': peer,
      'presented_registry_root': registry,
      'disposition': null,
      'rejected_branch': null,
    });
    s['status'] = 'MANUAL_REQUIRED';
    await _store.write(s);
    _viewFail(reason);
  }

  Future<Map<String, Object?>> evidence() => _run(() async {
    final s = await _load(), cases = s['cases']! as List;
    final c = cases.isEmpty
        ? {'local': s['views'], 'peer': []}
        : _object(cases.last);
    return {'local': c['local'], 'peer': c['peer']};
  });
  Future<Map<String, Object?>> recovery() => _run(() async {
    final s = await _load(),
        cases = s['cases']! as List,
        c = cases.isEmpty ? null : _object(cases.last);
    List<Map<String, Object?>> metadata(Object? v) => _views(v)
        .map(
          (v) => {
            for (final k in ['sequence', 'root', 'registry_root', 'key_epoch'])
              k: v[k],
          },
        )
        .toList();
    return {
      'status': s['status'],
      'reason': c?['reason'],
      'local': metadata(c?['local'] ?? s['views']),
      'peer': metadata(c?['peer'] ?? []),
      'disposition': c?['disposition'],
      'rejected_branch': c?['rejected_branch'],
      'presented_registry_root': c?['presented_registry_root'],
    };
  });
  Future<String> resolve(
    String disposition,
    String localRoot,
    String peerRoot,
  ) => _run(() async {
    final s = await _load();
    if (s['status'] != 'MANUAL_REQUIRED' ||
        ![
          'retain_accepted',
          'select_peer',
          'keep_blocked',
        ].contains(disposition)) {
      _viewFail();
    }
    final cases = s['cases']! as List,
        c = _object(cases.last),
        local = _views(c['local']),
        peer = _views(c['peer']);
    if (localRoot != (local.isEmpty ? _zeroRoot : local.last['root']) ||
        peerRoot != (peer.isEmpty ? _zeroRoot : peer.last['root'])) {
      _viewFail();
    }
    final safe =
        c['reason'] == 'ATLAS_ROLLBACK_REJECTED' &&
        peer.isNotEmpty &&
        peer.every(
          (v) =>
              (v['sequence']! as int) <= local.length &&
              v['root'] == local[(v['sequence']! as int) - 1]['root'],
        );
    c['disposition'] = disposition;
    c['rejected_branch'] = {
      'retain_accepted': 'peer',
      'select_peer': 'local',
      'keep_blocked': null,
    }[disposition];
    cases[cases.length - 1] = c;
    s['status'] = disposition == 'retain_accepted' && safe
        ? 'ACTIVE'
        : 'RECOVERY_PENDING';
    await _store.write(s);
    return s['status']! as String;
  });
  Future<int> compareEvidence(List<Map<String, Object?>> rawPeer) {
    final copied = rawPeer
        .take(_viewLimit + 1)
        .map((v) => Map<String, Object?>.of(v))
        .toList();
    return _run(() async {
      final s = await _load();
      await _active(s);
      var signed = <Map<String, Object?>>[];
      try {
        return await _checked(() async {
          if (copied.length > _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
          final bridge = await _bridge(s);
          final authorities = {
            _context['key_epoch']: _public,
            for (final p in bridge)
              _object(p['plan'])['new_epoch']: _bridgePublic(p),
          };
          for (final v in copied) {
            signed.add(
              await AtlasVaultAuthenticatedStateView._verify(
                v,
                authorities[v['key_epoch']] ?? _public,
              ),
            );
          }
          final peer = await _chain(signed, bridge), local = _views(s['views']);
          if (peer.isEmpty || local.isEmpty) {
            _viewFail('ATLAS_CHECKPOINT_REQUIRED');
          }
          final common = local.length < peer.length
              ? local.length
              : peer.length;
          for (var i = 0; i < common; i++) {
            if (local[i]['root'] != peer[i]['root']) {
              _viewFail('ATLAS_STATE_EQUIVOCATION');
            }
          }
          return common;
        });
      } on AtlasVaultStateViewException catch (e) {
        return _alarm(s, e.code, signed);
      }
    });
  }

  Future<bool> ingest(
    Map<String, Object?> rawView,
    List<Map<String, Object?>> rawRegistry,
    Map<String, Object?> rawCollection,
    Uint8List rawState,
  ) {
    final view = Map<String, Object?>.of(rawView),
        collection = Map<String, Object?>.of(rawCollection),
        registry = rawRegistry
            .take(_viewLimit + 1)
            .map((r) => Map<String, Object?>.of(r))
            .toList();
    final oversized = rawState.length > 1024 * 1024,
        bytes = oversized ? Uint8List(0) : Uint8List.fromList(rawState);
    return _run(() async {
      final s = await _load();
      await _active(s);
      var peer = <Map<String, Object?>>[];
      String? registryDigest;
      try {
        final duplicate = await _checked(() async {
          final bridges = await _bridge(s);
          final bridge = bridges.isEmpty ? null : bridges.last;
          final epoch = bridge == null
              ? _context['key_epoch']
              : _object(bridge['plan'])['new_epoch'];
          final public = bridge == null ? _public : _bridgePublic(bridge);
          final v = await AtlasVaultAuthenticatedStateView._verify(
            view,
            public,
          );
          peer = [v];
          if (['account_id', 'vault_id'].any((k) => v[k] != _context[k]) ||
              v['key_epoch'] != epoch) {
            _viewFail();
          }
          registryDigest = bridge == null
              ? AtlasVaultAuthenticatedStateView.registryRoot(registry)
              : AtlasVaultRevocation.registryRoot(registry);
          if (v['registry_root'] != registryDigest) {
            _viewFail('ATLAS_REGISTRY_SUBSTITUTION');
          }
          if (oversized) _viewFail('ATLAS_HISTORY_LIMIT');
          final c = AtlasVaultSignedStateCommitment.fromJson(collection);
          if (c.collectionId != _context['collection_id'] ||
              c.sequence != v['sequence'] ||
              c.root != v['collection_root'] ||
              c._value['state_sha256'] != _commitmentDigest(bytes)) {
            _viewFail();
          }
          if (!await Ed25519().verify(
            _rootSignatureMessage(c.root),
            signature: Signature(
              _base64(c._value['signature_b64'], exactLength: 64),
              publicKey: SimplePublicKey(public, type: KeyPairType.ed25519),
            ),
          )) {
            _viewFail();
          }
          final n = v['sequence']! as int, views = _views(s['views']);
          if (n <= views.length && v['root'] != views[n - 1]['root']) {
            _viewFail('ATLAS_STATE_EQUIVOCATION');
          }
          if (n < views.length) _viewFail('ATLAS_ROLLBACK_REJECTED');
          if (n == views.length) return true;
          if (n > _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
          final previous = views.isEmpty ? null : views.last;
          if (n != views.length + 1 ||
              v['previous_root'] != (previous?['root'] ?? _zeroRoot) ||
              v['previous_registry_root'] !=
                  (previous?['registry_root'] ?? _emptyRegistryRoot) ||
              c.previousRoot != (previous?['collection_root'] ?? _zeroRoot)) {
            _viewFail();
          }
          final body = _object(jsonDecode(utf8.decode(bytes)));
          _exact(body, {'format', 'version', 'route', 'records'});
          if (body['format'] != 'atlasvault-guarded-collection' ||
              body['version'] is! int ||
              body['version'] != 1 ||
              !['patch', 'snapshot', 'compaction'].contains(body['route'])) {
            _viewFail();
          }
          final raw = body['records']! as List;
          if (raw.length > _viewLimit) _viewFail('ATLAS_HISTORY_LIMIT');
          final records = <String, Object?>{};
          for (final item in raw) {
            final r = AtlasVaultOpaqueCiphertextEnvelope.fromJson(
              _object(item),
            );
            if (r.version != 1 ||
                records.containsKey(r.objectId) ||
                (r.keyEpoch != v['key_epoch'] &&
                    (_object(s['records'])[r.objectId] == null ||
                        _object(
                              _object(s['records'])[r.objectId],
                            )['envelope_sha256'] !=
                            _sha256Hex(_canonicalJsonBytes(r.toJson()))))) {
              _viewFail();
            }
            records[r.objectId] = {
              'object_id': r.objectId,
              'revision': r.revision,
              'content_sha256': r.contentSha256,
              'envelope_sha256': _sha256Hex(_canonicalJsonBytes(r.toJson())),
              'tombstone': r.tombstone,
            };
          }
          for (final entry in _object(s['records']).entries) {
            final old = _object(entry.value), candidate = records[entry.key];
            if (old['tombstone'] == true &&
                jsonEncode(_canonicalValue(old)) !=
                    jsonEncode(_canonicalValue(candidate))) {
              _viewFail('ATLAS_TOMBSTONE_RESURRECTION');
            }
            if (candidate == null) _viewFail('ATLAS_STALE_STATE');
          }
          s['records'] = records;
          s['views'] = [...views, v];
          return false;
        });
        if (duplicate) return false;
      } on AtlasVaultStateViewException catch (e) {
        return _alarm(s, e.code, peer, registryDigest);
      }
      await _store.write(s);
      return true;
    });
  }
}
